New fee estimation code

morcos opened the pull request:
I reworked the fee and priority estimation code to gather statistics by binning transactions of various fee rates and tracking how many blocks they took to get confirmed.   See [post](http://www.mail-archive.com/bitcoin-development@lists.sourceforge.net/msg06405.html) to mailing list.

morcos committed:
Create new BlockPolicyEstimator for fee estimates

This class groups transactions that have been confirmed in blocks into buckets, based on either their fee or their priority.  Then for each bucket, the class calculates what percentage of the transactions were confirmed within various numbers of blocks.  It does this by keeping an exponentially decaying moving history for each bucket and confirm block count of the percentage of transactions in that bucket that were confirmed within that number of blocks.

-Eliminate txs which didn't have all inputs available at entry from fee/pri calcs

-Add dynamic breakpoints and tracking of confirmation delays in mempool transactions

-Remove old CMinerPolicyEstimator and CBlockAverage code

-New smartfees.py

-Pass a flag to the estimation code, using IsInitialBlockDownload as a proxy for when we are still catching up and we shouldn't be counting how many blocks it takes for transactions to be included.

-Add a policyestimator unit test

morcos commented:
@gavinandresen, I've rebased this and added a couple of improvements, most notably, doing some tracking of transactions that are in the mempool and haven't been mined yet.  So in particular the are code changes from what you reviewed before.

gavinandresen commented:
My compiler is unhappy.
Compiler:
```
Apple LLVM version 6.0 (clang-600.0.57) (based on LLVM 3.5svn)
```
Unhappiness:
```
./streams.h:406:9: error: call to 'Serialize' is ambiguous
        ::Serialize(*this, obj, nType, nVersion);
        ^~~~~~~~~~~
policyestimator.cpp:175:13: note: in instantiation of function template specialization 'CAutoFile::operator<<<unsigned long>' requested here
    fileout << confAvg.size();
            ^
```

... and:

```
./serialize.h:555:6: error: member reference base type 'unsigned long' is not a structure or union
    a.Unserialize(is, (int)nType, nVersion);
    ~^~~~~~~~~~~~
./streams.h:416:11: note: in instantiation of function template specialization 'Unserialize<CAutoFile, unsigned long>' requested here
        ::Unserialize(*this, obj, nType, nVersion);
          ^
policyestimator.cpp:197:12: note: in instantiation of function template specialization 'CAutoFile::operator>><unsigned long>' requested here
    filein >> maxConfirms;
           ^
```

gavinandresen commented:
Since the format of fee_estimates.dat will change, CTxMemPool::WriteFeeEstimates should change:
```
fileout << 99900; // version required to read: 0.9.99 or later
 .... should be
fileout << 109900; // version required to read: 0.10.99 or later
```

morcos commented:
ah sorry, pushed too fast, i'd noticed that compilation problem, not sure why it was a problem, but i think i just pushed a fix...   will wait for more comments before adding more fixes.

gavinandresen commented:
Been running this at home for a couple of days, estimates look reasonable.

I've updated the bitcoind on core2.bitcoincore.org; as soon as it catches up with the chain it will start generating new fee/priority estimates again.

morcos commented:
Any thoughts on whether this is something we'd like to merge?  I'd like to get it in early enough in the 0.11 release cycle that it gets some usage.  I realize its a lot of complicated code, but it really only affects fee and priority estimates.  I think the increased accuracy of the estimates makes it worthwhile.  Compare (core2.) and bitcoincore.org/smartfee/fee_graph.html.

gavinandresen commented:
I'd like to see this in the 0.11 release.

jtimon commented:
Could you rename src/policyestimator.o to src/policy/estimator.o ?
I was planning on doing such a code movement and this seems like a perfect opportunity.
A later step would be to take the policy estimator out of the txmempool object, but this movement already helps with that independently of what you're doing inside the the estimator itself which I haven't reviewed.
This is what I had in mind after the movement: https://github.com/jtimon/bitcoin/commit/ba88aa141985bfba94a67289cf176e3a562971da

It will obviously have to change after this PR, but maybe you can get some ideas from there as well, and hopefully make the "rebase/rewrite" of that encapsulation easier. If you manage to incorporate the essence of that commit in your PR, I would love you for it, but that's probably too much to ask at this point.
In any case, I would already be happy with the file rename.

morcos commented:
I'd be happy to give it shot.  I'll ping you when I get around to it.

morcos commented:
Rebased, cleaned up and squashed original commits.

Added 3 new commits.
This is ready for more review.
@jtimon I made the move you asked for but held off for now on the other changes until I know what the right way to couple the mempool to the estimator is.  There are now 3 interface points instead of the previous 1 that the old code had.

jtimon commented:
I had more time to review this.
I don't understand what you mean by the 3 interface points but BlockPolicyEstimator is already decoupled from CTxMemPool with your current code.
You made the mistake of coupling CTxMemPoolEntry with CTxMemPool in https://github.com/morcos/bitcoin/commit/ff34d782517f9cabed87772db96a7e7a76ac5db0
and although you fix it in https://github.com/morcos/bitcoin/commit/64bf2bbe4400d1c4da23f56ebb8dee595f5055fe that mistake should never reach the history. I'm not sure if you planned to squash that commit or leave it separate.

So to decouple BlockPolicyEstimator from txmempool.o you just need to also take AllowFreeThreshold and AllowFree with you, and separate CTxMemPoolEntry to a different file.
At that point, maybe something like policy/fees makes more sense than policy/estimator, we can put more fee-related stuff there later (like CFeeRate, once CTxOut gets decoupled from it).

Also, why couple CBlockPolicyEstimator to the minRelayTxFee global when it can take CTxMemPool::minRelayFee (which is only used to pass it to CBlockPolicyEstimator::Read which ends up not using it) ??

You have some good opportunities to reduce dependency on boost or at least not increase it (I'm talking about the foreach loops), but if you don't want to do it you can at least explicitly include the dependency in the new file.

I have created a branch with my many suggestions in form of code:

https://github.com/jtimon/bitcoin/commits/morcos_new-estimates

The last commit doesn't pass the tests because mempool.HasNoInputsOf(tx) is always called when creating a new entry: CTxMemPoolEntry::hadNoDependencies is never set to false by default.
The rest of the commits compile individually although some of them would increase your total diff to master more than others (some are "free" in that sense, like the file rename one) so it's up which one you include.
Feel free to squash, edit, reword, kill or reorder the commits as you think it's best.

morcos commented:
Thanks @jtimon.  I included several of your commits in an all purpose SQUASHME fixup.  I also changed my move commit to yours.  My original commit and 2 recent additions are unchanged.   I think we can PR your other changes separately afterwards.  I have them rebased on top of this here:
https://github.com/morcos/bitcoin/commits/new-estimates_jtimon

@gavinandresen, if you wouldn't mind reviewing the new commits.  The first commit should be what you reviewed already (rebased and tidied a little).

morcos commented on src/txmempool.cpp:
```diff
@@ -419,20 +92,20 @@ void CTxMemPool::AddTransactionsUpdated(unsigned int n)
-        nTransactionsUpdated++;
-        totalTxSize += entry.GetTxSize();
-    }
+    mapTx[hash] = entry;
```
The transaction being added to the mempool has actually been copied twice at this point.  The first time when the CTxMemPoolEntry was created, and now again when its copied into the map, which will be its permanent home.  This was not introduced by this PR, but a further PR should consider eliminating one of the copy operations.

jtimon commented:
Thanks for incorporating some of my suggestions, I'm specially happy about not using the minRelayTxFee global inside the estimator which was the most important nit. I agree all the suggestions not incorporated can wait and it's better not to slow this down.

sdaftuar commented on src/txmempool.cpp:
```diff
@@ -385,7 +58,7 @@ CTxMemPool::CTxMemPool(const CFeeRate& _minRelayFee) :
     // Confirmation times for very-low-fee transactions that take more
     // than an hour or three to confirm are highly variable.
-    minerPolicyEstimator = new CMinerPolicyEstimator(25);
+    minerPolicyEstimator = new CBlockPolicyEstimator(_minRelayFee);
```
The comment above is no longer relevant here.

sdaftuar commented on src/txmempool.h:
```diff
@@ -30,6 +30,8 @@ inline bool AllowFree(double dPriority)
 /** Fake height value used in CCoins to signify they are only in the memory pool (since 0.8) */
 static const unsigned int MEMPOOL_HEIGHT = 0x7FFFFFFF;
 
+class CTxMemPool;
```
nit: this forward declaration doesn't appear to be necessary

morcos commented:
Addressed nits (above and offline) by @sdaftuar and improved the readability of the rpc test.
@gavinandresen, I know there are a lot of small added commits here, and I'd like to squash them all down but just wanted to be sure you saw the changes in 64bf2bb.  This slightly changes the startup behavior, but otherwise all other changes we just cleanups, and no change to behavior.

sdaftuar commented on src/policy/fees.cpp:
```diff
@@ -0,0 +1,529 @@
+        }
+    }
+
+    LogPrint("estimatefee", "%3d: For conf succeess %s %4.2f need %s %s: %12.5g from buckets %8g - %8g  Cur Bucket stats %6.2f%%  %8.1f/(%.1f+%d mempool)\n",
```
nit: succeess -> success

sdaftuar commented:
I've reviewed the code (particularly to understand the details of the calculations in policy/fees.cpp), and I've tested by running a node with this code, and this looks good to me (and improves the estimates substantially).  ACK.

jtimon commented:
I'm going to make a summary of my review of this PR, hopefully it helps getting it merged soon.
Although, I have not reviewed or tested the changes in the estimator itself, I have paid special attention on making sure that it only changed fee policy behavior, so I'm convinced that this is completely consensus and network safe, anything-non-fee-estimator-related safe.
Reaching this conclusion was not hard because although coupled with txmempool, the estimator is encapsulated.
Apart from the changes in the estimator, this is an improvement in modularity, and will open the door to further improvements in fee policy encapsulation (the min relay fee should stop being a global, all functions using it should be methods of the estimator class, but one step at a time).
I repeated the moveonly part of this PR manually so if it had been separated in an individual commit I could have easily verified it, but then I git pruned that (I should have kept it).

In summary, still-untested-but-pretty-confident-ACK.

Diapolo commented:
I didn't try to check or understand that code, but wanted to ask if the GUI (coincontrol smart fees) is working with this without problems or if that needs to be verified.

morcos commented:
@Diapolo Thanks for thinking about that.  I think it could certainly use some more testing.  There are two things in my mind that could use improving.
1) I think the default txConfirmTarget should be greater than 1.  There are periods of time when 1 can be quite a high fee temporarily.  I plan to open this discussion in a different PR at some point, because its a bit of a separate discussion.  This doesn't really affect the GUI, though since that uses a default of 25.
2) It's also possible at times to not have an answer for `estimatefee 1` but to have answer for all other estimates.  This is a feature not a bug, but I think it would be possible to design a better UI experience in that case.  Perhaps the slider bounces back to the first legitimate estimate, or gives you a helpful message such as "No amount of fee will get you reliably confirmed in the next block".

bitcoinfees commented:
>  There are periods of time when 1 can be quite a high fee temporarily

A suggestion: if fractional arguments are allowed for `estimatefee`, it could avoid the need for `MIN_SUCCESS_PCT`, which in my view discards useful information, and perhaps is the reason for these kinds of problems (e.g. a fee bucket which for a given `confTarget` has a `curPct` that is close to the 0.85 threshold would result in unstable behavior, if I've understood properly).

With fractional `confTarget`, each fee bucket could be represented by its mean conf, and compared to `confTarget` directly to determine success. The default `txConfirmTarget` could then be set to fractional values, e.g. 1.2, which would perhaps result in greater stability.

bitcoinfees commented:
> not have an answer for estimatefee 1

With regard to fee buckets clearing the `MIN_SUCCESS_PCT` hurdle of 0.85, I've found in my own studies that the main reason why transactions with relatively high fees don't make it into the next block is because of the interval between miners updating their block transaction lists, which I believe is on the order of 30-60 seconds ([source](http://bitcoin.stackexchange.com/questions/32892/how-often-do-miners-update-their-block-transaction-list)). 

This means that if a transaction has been in the mempool for less than 60 seconds, there is a chance it won't be considered for block inclusion at all by the miner, regardless of the tx fee. After removing these transactions from the estimation, I've found that the relative frequency of high-fee transactions being included in the next block is pretty high and consistent, typically > 95%.

morcos commented:
@bitcoinfees, thanks for the comments.  I think asking for the mean number of blocks to confirmation is a slightly different question than I answered.  It's not clear if one question is better than another, but the way I defined the question is what fee would you need to have a sufficiently high percentage chance of getting confirmed in the desired number of blocks, so it kind of requires a MIN_SUCCESS_PCT.  But to your point, I do think that there could be a lot of improvements to this estimation algorithm, and hopefully they could build off the framework it provides.

I agree with your conclusion on why the success percentage has a ceiling around 90%.  Hopefully with an improved transaction selection algorithm for mining in the future that will get better.

gavinandresen commented:
Code review and lightly tested ACK.

I'm updating core2.bitcoincore.org to run this version of the code now.

morcos commented:
squashed to hopefully make it mergeable

fanquake commented on src/policy/fees.cpp:
```diff
@@ -0,0 +1,529 @@
+// Copyright (c) 2009-2010 Satoshi Nakamoto
+// Copyright (c) 2009-2015 The Bitcoin developers
+// Distributed under the MIT/X11 software license, see the accompanying
```
nit: just MIT

fanquake commented on src/policy/fees.h:
```diff
@@ -0,0 +1,276 @@
+// Copyright (c) 2009-2010 Satoshi Nakamoto
+// Copyright (c) 2009-2015 The Bitcoin developers
+// Distributed under the MIT/X11 software license, see the accompanying
```
nit: also here

morcos commented:
nits addressed

laanwj commented:
utACK
